### ISSUE TITLE
MISC_LUA does not work on boxes w/o 5 Ghz

### ISSUE THREAD

**Issue by uahim** [jhubig/FritzBoxShell#42]
MISC_LUA did not work at all for me - I assume that the format is different on boxes that only have 2,4 GHz. in the output on my box, "LAN" is uppercase and there is no "desc", only "type":"WLAN - 2,4 GHz"

fixed it for myself with an additional
`totalConnectionsWLAN2G=$(grep -io '"type":"wlan - 2' <<< $overview | wc -l)`
(but I don't know if "type":"wlan - (...)" is present on boxes *with* 5ghz)

and an exta -i in grep here
`totalConnectionsLAN=$(echo $overview | grep -iow '"type":"lan"' | wc -l)`

also, is it just me or is MISC_LUA very, very slow? at least on my box, it takes up to 30 seconds or so.

**Comment by jhubig** [comment]
Hello,

I have created a new function LANcount (through TR064 instead of LUA) which you can try with

`fritzBoxShell.sh LAN COUNT`

Please download the latest code here: https://github.com/jhubig/FritzBoxShell/archive/refs/heads/master.zip (Don't forget to set again your config inside FritzBoxShellConfig.sh if needed).

Let me know if this works better for you. Thanks.

Cheers,
Johannes

**Comment by Ldwg2002** [comment]
Dear Johannes,

i was curious about it and did some testing. **LAN COUNT works fine**, even for my router (7490 with 25 clients). The response time corresponds to the load of the FritzBox resp. number of (registered or) active clients. It is faster with less clients and less load, measured with time: 3 sec. with 3 clients, up to 10-20 sec. with 25 clients and 70 % traffic load, another FB takes 5 sec. with 20 clients and no traffic load.

The very best, it works independently from the configuration (as Router, cascaded Router or as AccessPoint/IP-Client-Mode). 

I have to report: `a break condition, if the right answer is ZERO`, the subroutine reports
`user@rpi2bv11:~/poll $ /bin/bash ./fritzBoxShell.sh --boxip 10.42.1.1 --boxuser "fbInfo" --boxpw "fbPwd" LAN COUNT
./fritzBoxShell.sh: Zeile 651: /tmp/ethernet_count.tmp: Datei oder Verzeichnis nicht gefunden
Number of Ethernet connections:
rm: das Entfernen von '/tmp/ethernet_count.tmp' ist nicht möglich: Datei oder Verzeichnis nicht gefunden
`
All other AccessPoints have LAN-devices connected and reply correct with 1 or 2 within 2-3 sec.
Thanks a lot. I am happy to proceed CLI-monitoring with your FritzBoxShell script. :-) 

Ludwig

**Comment by jhubig** [comment]
Hi Ludwig,

thanks for your hints. i did another update: https://github.com/jhubig/FritzBoxShell/commit/de8a0dd99d43177b512b5ae23ed6489cfce70301

`/fritzBoxShell.sh LAN COUNT`
`NumberOfEthernetConnections: 16`

Bei keinen erkannten Hosts

`/fritzBoxShell.sh LAN COUNT`
`NumberOfEthernetConnections: 0`

**Comment by jhubig** [comment]
Another update was performed with more robust checking of number of hosts and/or ethernet connections (https://github.com/jhubig/FritzBoxShell/commit/0ef93c96991e9cea531c0270f3740f4855d5d943).

Please check. Use the latest code from here: https://github.com/jhubig/FritzBoxShell/archive/refs/heads/master.zip

Thanks.

**Comment by Weltkugel** [comment]
Thank you for FritzBoxShell.

I use the MISC_LUA function to detect the presence of cell phones. To do this, I search for network participants in the _$overview_ variable. This works very reliably.

`overview=$(wget -O - --post-data "xhr=1&sid=$SID&page=overview&xhrId=first&noMenuRef=1" "http://$BoxIP/data.lua" 2>/dev/null)`
...
`echo $overview | jq -r '.' | grep -c -e Pixel-9-Pro -e iPhone-SE`

Perhaps this is an idea for another script function.

**Comment by uahim** [comment]
> Another update was performed with more robust checking of number of hosts and/or ethernet connections ([0ef93c9](https://github.com/jhubig/FritzBoxShell/commit/0ef93c96991e9cea531c0270f3740f4855d5d943)).
> 
> Please check. Use the latest code from here: https://github.com/jhubig/FritzBoxShell/archive/refs/heads/master.zip
> 
> Thanks.

thanks but GetHostNumberOfEntries is not available on older boxes, I'll just stick with the slow method

**Comment by Ldwg2002** [comment]
Did you notice that you need to install the XML tool? On RaspberryPi `sudo apt install xmlstarlet` and everything is working fine for me now.

**Comment by jhubig** [comment]
Maybe also try the today added function get_filtered_clients().

2.4 for 2.4 Ghz Network clients
5 for 5 Ghz Network clients
ETH for ethernet Network clients
all for all Network clients

`/fritzBoxShell.sh COUNT 5`

**Comment by uahim** [comment]
> `/fritzBoxShell.sh  COUNT all -withIP`

nice, that's a good idea

can the code be changed so it only gets the ip when actually needed? e.g. by moving ` if [ "$show_ip" == "-withIP" ];` from get_ip_from_mac to get_filtered_clients ?

because I ran into a problem that has actually nothing to do with your code per se, but is closely connected to "-withIP": when running on cygwin, even without "-withIP" jq throws errors. the issue seems to be the dollar sign of the variable here: `mac=$(echo "$unique_clients" | jq -r ".[$i].mac")`, it throws `jq: error: syntax error, unexpected INVALID_CHARACTER (Windows cmd shell quoting issues?) at <top-level>`
and when I put the statement in single instead of double quotes, it says `$i is not defined at top-level `.
with a hardcoded ".[0].mac" for debugging purposes, it works fine when used without "-withIP". *with* "-withIP" of course it fails a few lines further down
```
error: syntax error, unexpected INVALID_CHARACTER (Windows cmd shell quoting issues?) at <top-level>, line 1:
].ip = "192.168.178.33"
```
which again seems to be because of the dollar sign of the i variable.

I assume one could try to somehow concat strings to get around the var=$(expr)/loop problem/win shell limitation, but have no idea how to do that.

thanks

**Comment by jhubig** [comment]
> can the code be changed so it only gets the ip when actually needed? e.g. by moving ` if [ "$show_ip" == "-withIP" ];` from get_ip_from_mac to get_filtered_clients ?

You can just call it without the -withIP

`/fritzBoxShell.sh  COUNT all`

For the cygwin topic I changed the code and commited (https://github.com/jhubig/FritzBoxShell/commit/b049495844f8528ed892bcdac513da830c1312ad) a new version. Could you please check? Thanks.

**Comment by uahim** [comment]
> > can the code be changed so it only gets the ip when actually needed? e.g. by moving ` if [ "$show_ip" == "-withIP" ];` from get_ip_from_mac to get_filtered_clients ?
> 
> You can just call it without the -withIP

yes, sorry if that wasn't clear - the jq error was thrown with or without -withIP

> `/fritzBoxShell.sh COUNT all`
> 
> For the cygwin topic I changed the code and commited ([b049495](https://github.com/jhubig/FritzBoxShell/commit/b049495844f8528ed892bcdac513da830c1312ad)) a new version. 

this works like a charm on cywgin, thank you!

**Comment by jhubig** [comment]
@uahim Could you please try with the latest commit and check if the jq error out of cygwin is still thrown? Thanks.
